homeassistant/components/androidtv/__init__.py

```diff
@@ -1 +1,195 @@
+import json
+import logging
+import os
+
+from adb_shell.auth.keygen import keygen
+from androidtv import state_detection_rules_validator
+from androidtv.adb_manager.adb_manager_sync import ADBPythonSync
+from androidtv.setup_async import setup
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_DEVICE_CLASS,
+    CONF_HOST,
+    CONF_PORT,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.storage import STORAGE_DIR
+
+from .const import (
+    ANDROID_DEV,
+    ANDROID_DEV_OPT,
+    CONF_ADB_SERVER_IP,
+    CONF_ADB_SERVER_PORT,
+    CONF_ADBKEY,
+    CONF_STATE_DETECTION_RULES,
+    DEVICE_ANDROIDTV,
+    DEVICE_FIRETV,
+    DOMAIN,
+    MIGRATION_DATA,
+    SIGNAL_CONFIG_ENTITY,
+)
+
+PLATFORMS = ["media_player"]
+RELOAD_OPTIONS = [CONF_STATE_DETECTION_RULES]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _setup_androidtv(hass, config):
+    """Generate an ADB key (if needed) and load it."""
+    adbkey = config.get(CONF_ADBKEY, hass.config.path(STORAGE_DIR, "androidtv_adbkey"))
+    if CONF_ADB_SERVER_IP not in config:
+        # Use "adb_shell" (Python ADB implementation)
+        if not os.path.isfile(adbkey):
+            # Generate ADB key files
+            keygen(adbkey)
+
+        # Load the ADB key
+        signer = ADBPythonSync.load_adbkey(adbkey)
+        adb_log = f"using Python ADB implementation with adbkey='{adbkey}'"
+
+    else:
+        # Use "pure-python-adb" (communicate with ADB server)
+        signer = None
+        adb_log = f"using ADB server at {config[CONF_ADB_SERVER_IP]}:{config[CONF_ADB_SERVER_PORT]}"
+
+    return adbkey, signer, adb_log
+
+
+def validate_state_det_rules(state_det_rules):
+    """Validate a string that contain state detection rules and return a dict."""
+    if not state_det_rules:
+        return None
+
+    try:
+        json_rules = json.loads(state_det_rules)
+    except ValueError:
+        _LOGGER.warning("Error loading state detection rules")
+        return None
+    try:
+        state_detection_rules_validator(json_rules, ValueError)
+    except ValueError as exc:
+        _LOGGER.warning("Invalid state detection rules: %s", exc)
+        return None
+    return json_rules
+
+
+async def async_connect_androidtv(
+    hass, config, *, state_detection_rules=None, timeout=10.0
+):
+    """Connect to Android device."""
+    address = f"{config[CONF_HOST]}:{config[CONF_PORT]}"
+
+    adbkey, signer, adb_log = await hass.async_add_executor_job(
+        _setup_androidtv, hass, config
+    )
+
+    aftv = await setup(
+        config[CONF_HOST],
+        config[CONF_PORT],
+        adbkey,
+        config.get(CONF_ADB_SERVER_IP, ""),
+        config[CONF_ADB_SERVER_PORT],
+        state_detection_rules,
+        config[CONF_DEVICE_CLASS],
```

Comment:
```suggestion
        config.get(CONF_ADB_SERVER_IP),
        config.get(CONF_ADB_SERVER_PORT),
        state_detection_rules,
        config.get(CONF_DEVICE_CLASS),
```

Reply:
I don't think that we can avoid breaking change. The problem is that the entity is currently using an unique ID based on serial No, because we are going to change unique ID (based on mac) a new entity with new entity ID will be created.

Reply:
I know very little about config flow stuff, but @balloob previously said ([link](https://github.com/home-assistant/core/issues/35325#issuecomment-625920341)):

> It doesn't have to be a breaking change because we can do a migration. `async_migrate_entries` exists in `helpers/entity_registry.py`.

I haven't looked at @tkdrob's suggested changes, so for all I know this might be what he's doing.

Reply:
I don't see any migration here, this seems more related to simplify config flow. I start merging this changes, we can come back to breaking change later.